app/controllers/reviews_controller.rb

```diff
@@ -19,14 +19,28 @@ def create
     if Review.where(user_id: @user.id, book_id: @book.id).any?
       flash[:notice] = "You've already written a review for this book."
       redirect_to book_path(@book)
-
     elsif @review.save
       flash[:success] = "Review added successfully!"
       redirect_to book_path(@book)
       ReviewMailer.new_review(@review).deliver_later
     else
-      flash[:warning] = @review.errors.full_messages.join(', ')
-      render :new
+      create_or_update_failure
+    end
+  end
+
+  def edit
+    session[:return_to] ||= request.referer
+    @review = Review.find(params[:id])
+    @ratings_collection = RATINGS
+  end
+
+  def update
+    @review = Review.find(params[:id])
+    if @review.update(review_params)
+      flash[:success] = "Review udpated successfully!"
+      redirect_to session[:return_to]
+    else
+      create_or_update_failure
```

Comment:
Nice pattern to abstract logic into a private method. That said, you don't really need to here - or if you were going to do it, consider abstracting a bigger chunk of code. You can leave it though, just wanted to give you some perspective.